src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +60,206 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
     return bf_response_new_success(response, NULL, 0);
 }
 
+static void bf_nop_free(struct bf_chain **chain)
+{
+    UNUSED(chain);
+}
+
+static size_t _bf_cli_num_rules(bf_list *chain_list)
+{
+    bf_assert(chain_list);
+
+    size_t count = 0;
+    bf_list_foreach (chain_list, chain_node) {
+        struct bf_chain *chain = bf_list_node_get_data(chain_node);
+
+        count += bf_list_size(&chain->rules);
+    }
+
+    return count;
+}
+
+static int _bf_get_ctr_vals(bf_list *chain_list, struct bf_counter *counters)
+{
+    int counter_index = 0;
+    int r;
+
+    bf_assert(chain_list);
+    bf_assert(counters);
+
+    bf_list_foreach (chain_list, chain_node) {
+        struct bf_chain *chain = bf_list_node_get_data(chain_node);
+        if (!chain)
+            bf_err("Chain list element pointed to null data\n");
+
+        bf_assert(chain);
+
+        struct bf_cgen *cgen = bf_ctx_get_cgen(chain->hook, NULL);
+        if (!cgen)
+            bf_err("Got a null cgen for chain element\n");
+
+        // Iterate over each rule in the current chain
+        int map_idx = 0;
+        bf_list_foreach (&chain->rules, rule_node) {
+            struct bf_rule *rule = bf_list_node_get_data(rule_node);
+            if (!chain)
+                bf_err("Rule list element pointed to null data\n");
+
+            if (rule->counters) {
+                // Query the BPF map for the counter values
+                r = bf_cgen_get_counter(cgen, map_idx,
+                                        &counters[counter_index]);
+                if (r < 0)
+                    return bf_err_r(r, "Failed to get rule counter\n");
+            }
+
+            /* Note: The map has entries for every rule regardless of whether the
+             * input rule specified "COUNTERS" or not. Thus, advancing map_index
+             * is unconditional. When rule->counters == false, the corresponding
+             * counter value will remain {0,0}. */
+            counter_index++;
+            map_idx++;
+        }
+
+        // Chain's policy counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_POLICY,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get policy counter\n");
+        counter_index++;
+
+        // Chain's error counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_ERRORS,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get error counter\n");
+    }
+
+    return 0;
+}
```

Comment:
I've published comments on this function, but it will likely be removed according to my comments below.